File: ios/RDRCTStoryView.m
//
//  RDRCTStoryView.m
//  react-native-related-digital
//
//  Created by Egemen Gülkılık on 22.08.2023.
//


#import "RDRCTStoryView.h"
#import "react_native_related_digital-Swift.h"

@interface RDRCTStoryView()
@end

NSString *const RDRCTStoryViewActionIDKey = @"actionId";
int const RDRCTStoryViewTag = 1903;

@implementation RDRCTStoryView

- (instancetype)init {
  self = [super initWithFrame:CGRectZero];
  return self;
}

- (void)setActionID:(NSString *)actionId {
  _actionId = actionId;

  dispatch_async(dispatch_get_main_queue(), ^{
    [self layoutSubviews];
  });
}

- (void)layoutSubviews {
  [self setupView];
}

- (void)setupView {
  if ([self viewWithTag:RDRCTStoryViewTag]) {
    [[self viewWithTag:RDRCTStoryViewTag] removeFromSuperview];
  }

  [RDRCTHelper.shared getStoryView:nil
                   withUrlDelegate:nil
             withCompletionHandler:^(RDRCTStoryView *storyView) {
      if(storyView != nil) {
          
      }
      
      
             }];

    /*
  let storyView : VisilabsStoryHomeView if (actionId != nil && actionId != "") {
    storyView = Visilabs.callAPI().getStoryView(actionId : Int(actionId !), urlDelegate : self)
  }
  else {storyView = Visilabs.callAPI().getStoryView(actionId
                                                    : nil, urlDelegate
                                                    : self)}

  storyView.tag = RDStoryView
                      .viewTag

                          self.addSubview(storyView) storyView.frame = self.frame
     */
}

- (void)close {
  NSString *actionId = self.actionId ?: @"";
  if (self.onItemClicked) {
    self.onItemClicked(@{RDRCTStoryViewActionIDKey : actionId});
  }
  self.actionId = nil;
}

@end

File: ios/RDRCTStoryView.swift
//
//  RDRCTStoryView.swift
//  react-native-related-digital
//
//  Created by Egemen Gülkılık on 23.08.2023.
//

import UIKit
import React
import RelatedDigitalIOS

private typealias NativeRD = RelatedDigitalIOS.RelatedDigital

@objc public class RDRCTStoryView: RCTView, RDStoryURLDelegate {
    
    static let viewTag = 1789
    
    static let storyItemUrlKey = "storyItemUrl"
    
    @objc var actionId: String? = nil
    
    @objc var onItemClicked: RCTBubblingEventBlock?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    public override func layoutSubviews() {
        setupView()
    }
    
    private func setupView() {
        let storyView = UIView()
        if self.viewWithTag(Self.viewTag) != nil {
            self.viewWithTag(Self.viewTag)?.removeFromSuperview()
        }
        
        NativeRD.getStoryViewAsync(actionId: Int(actionId ?? "") , urlDelegate: self) { storyHomeView in
            if let storyHomeView = storyHomeView {
                storyHomeView.translatesAutoresizingMaskIntoConstraints = false
                storyView.addSubview(storyHomeView as UIView)
                
                NSLayoutConstraint.activate([storyHomeView.topAnchor.constraint(equalTo: storyView.topAnchor),
                                             storyHomeView.bottomAnchor.constraint(equalTo: storyView.bottomAnchor),
                                             storyHomeView.leadingAnchor.constraint(equalTo: storyView.leadingAnchor),
                                             storyHomeView.trailingAnchor.constraint(equalTo: storyView.trailingAnchor)])
                
            }
        }
        
        storyView.tag = Self.viewTag
        
        self.addSubview(storyView)
        storyView.frame = self.frame
    }
    
    public func urlClicked(_ url: URL) {
        self.onItemClicked?([RDRCTStoryView.storyItemUrlKey: url.absoluteString])
    }
    
}
